### PR TITLE
 session, staking: introduce static limit on the number of validators

### LINKED ISSUE #9724: session, staking: introduce static limit on the number of validators
We need to introduce a static value for the maximum number of validators allowed in staking and session, we can then use this limit to replace a couple of storage entries with length-bounded data structures, e.g. `DisabledValidators`, `Invulnerables`, `OffendingValidators` can use `BoundedVec`.

See: https://github.com/paritytech/substrate/pull/9448#discussion_r693232481

### PR DESCRIPTION
**Pull request by Doordashcon**
Attempts to resolve #9724 

polkadot companion: https://github.com/paritytech/polkadot/pull/5905
cumulus companion: https://github.com/paritytech/cumulus/pull/1558

### CODE REVIEW

**Comment by Doordashcon** [comment]
@andresilva is this the general direction you envisioned?

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -495,6 +495,10 @@ impl_opaque_keys! {
 	}
 }
 
+parameter_types! {
+	pub const ValidatorSet: u32 = 3072;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:499]: `MaxValidators` makes more sense to me, WDYT?

File: frame/session/src/historical/mod.rs [hunk 2]
```diff
@@ -130,7 +130,7 @@ impl<T: Config> ValidatorSet<T::AccountId> for Pallet<T> {
 	}
 
 	fn validators() -> Vec<Self::ValidatorId> {
-		super::Pallet::<T>::validators()
+		super::Pallet::<T>::validators().to_vec()
```
Comment (kianenigma) [inline frame/session/src/historical/mod.rs:133]:
```suggestion
		super::Pallet::<T>::validators().into_inner()
```

File: frame/session/src/lib.rs [hunk 3]
```diff
@@ -412,6 +413,10 @@ pub mod pallet {
 
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
+
+		// Max number of validators
```
Comment (kianenigma) [inline frame/session/src/lib.rs:417]:
```suggestion
		/// Max number of validators.
```

File: frame/session/src/lib.rs [hunk 4]
```diff
@@ -466,6 +471,7 @@ pub mod pallet {
 					);
 					self.keys.iter().map(|x| x.1.clone()).collect()
 				});
+			let initial_validators_bounded: BoundedVec<T::ValidatorId, T::ValidatorSet> = initial_validators_0.clone().try_into().expect("Too many current validators");
```
Comment (kianenigma) [inline frame/session/src/lib.rs:474]:
```suggestion
			let initial_validators_bounded: BoundedVec<T::ValidatorId, T::ValidatorSet> = initial_validators_0.clone().try_into().expect("Too many initial validators");
```

File: frame/session/src/lib.rs [hunk 5]
```diff
@@ -647,7 +653,9 @@ impl<T: Config> Pallet<T> {
 		let session_keys = <QueuedKeys<T>>::get();
 		let validators =
 			session_keys.iter().map(|(validator, _)| validator.clone()).collect::<Vec<_>>();
-		Validators::<T>::put(&validators);
+
+		let validator2: BoundedVec<T::ValidatorId, T::ValidatorSet> = validators.clone().try_into().expect("Too many current validators");
```
Comment (kianenigma) [inline frame/session/src/lib.rs:657]: You certainly cannot and should not panic here -- need to handle it defensively, either use `WeakBoundedVec`, or truncate to fit (better).

**Review by kianenigma — CHANGES_REQUESTED** [review]
Good start, but needs some more work, looking forward!

File: bin/node/runtime/src/lib.rs (hunk 1 above)
Comment (Doordashcon) [reply bin/node/runtime/src/lib.rs:499]: LGTM!

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1743895

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1743975

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1744048

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - cargo-check-benches 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1744086

File: frame/session/src/lib.rs (hunk 5 above)
Comment (Doordashcon) [reply frame/session/src/lib.rs]: Resolved?

File: frame/session/src/lib.rs [hunk 6]
```diff
@@ -647,7 +655,10 @@ impl<T: Config> Pallet<T> {
 		let session_keys = <QueuedKeys<T>>::get();
 		let validators =
 			session_keys.iter().map(|(validator, _)| validator.clone()).collect::<Vec<_>>();
-		Validators::<T>::put(&validators);
+
+		let bounded_validator_set: WeakBoundedVec<T::ValidatorId, T::MaxValidators> =
+			validators.clone().try_into().expect("Max validators reached");
```
Comment (kianenigma) [inline frame/session/src/lib.rs:660]: this panic is still not valid.

File: frame/session/src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply frame/session/src/lib.rs]: not yet -- you cannot panic in any runtime code path such as on_initialize or a transaction.

**Review by kianenigma — CHANGES_REQUESTED** [review]
Still creating some panicing code that is not allowed. 

The correct way to to do this is actually to bound the interfaces linking staking and session pallet as well, something that is not done yet but is toyed with as a part of #11499 and alike.

For example, the session pallet should be able to ENSURE that its `SessionManager` does not return more than `x` validators. 

This is all likely to be part of a greater effort. We can continue trying it out here, but I can't guarantee that it will be easy 😅

File: frame/session/src/lib.rs (hunk 5 above)
Comment (Doordashcon) [reply frame/session/src/lib.rs]: Mapping to a custom error would be ideal?

**Pull request closed without merging**